### PR TITLE
Add support for custom aspect ratios

### PR DESCRIPTION
**Pull request by Filoppi**
This PR adds support for a user specified aspect ratio.

The main purpose of this is to allow users to play at specific game aspect ratios that don't match their screen.
For example, to play Super Mario Galaxy with the 21:9 cheat/mod, unless I had a 21:9 monitor, under normal circumstances I'd have to use the "Stretch" aspect ratio mode and size the window myself, by making a rough guess, withought having the possiblity of making the window fullscreen (black borders).
With this change it's now be possible to play (e.g.) games at 21:9 on both a 16:9 screen or a 32:9 one.

Another advantage this PR gives is that users can use the 1:1 aspect ratio, or the XFB native output aspect ratio (e.g. 640x528, pixel perfect nearest neighbor with no stretching). It's my understanding that this feature has been requested for a long time.

This PR is NOT related with the widescreen hack, though it works well with it, just like the "Stretch" aspect ratio mode.
All of the code surrounding the aspect ratio managment have been polished a little bit (mostly commenting and naming wise).

As a plus I threw some fixes for the "Auto" Internal resolution mode, which wasn't always calculated as good as it could have after my last PR (https://github.com/dolphin-emu/dolphin/pull/12000), and also it wasn't updated when resizing the window, making it unreliable.

Dolphin_IOg5SkIJ8r
Dolphin_GTh1Ieie5L

I suggest to review changes commit by commit, as they are mostly separate.

### CODE REVIEW

**Comment by MayImilae** [comment]
I would like for the Custom Aspect Ratio controls to only appear when Custom is selected. Other than that I think this is alright from a UX perspective.

**Comment by Filoppi** [comment]
> I would like for the Custom Aspect Ratio controls to only appear when Custom is selected. Other than that I think this is alright from a UX perspective.

They are greyed out if Custom AR is not selected. Do you want me to entirely hide them?

**Comment by MayImilae** [comment]
Yep. IMO hiding is better for this situation since custom aspect ratio setting are only useful when custom aspect ratio is active, so it's good to only reveal it when it is useful.

**Comment by Filoppi** [comment]
@MayImilae done

File: Source/Core/VideoCommon/Present.cpp [hunk 1]
```diff
@@ -201,20 +204,42 @@ void Presenter::ProcessFrameDumping(u64 ticks) const
...
+  OnBackbufferSet(size_changed, is_first);
+}
+
+void Presenter::OnBackbufferSet(bool size_changed, bool is_first_set)
+{
   UpdateDrawRectangle();
+
+  // Automatically update the resolution scale if the window size changed,
+  // or if the game XFB resolution changed.
+  if (size_changed && !is_first_set && g_ActiveConfig.iEFBScale == EFB_SCALE_AUTO_INTEGRAL &&
```
Comment (iwubcode) [inline Source/Core/VideoCommon/Present.cpp:238]: Please add braces here

File: Source/Core/VideoCommon/Present.cpp [hunk 2]
```diff
@@ -201,20 +204,42 @@ void Presenter::ProcessFrameDumping(u64 ticks) const
...
   m_backbuffer_width = backbuffer_width;
   m_backbuffer_height = backbuffer_height;
   UpdateDrawRectangle();
+
+  OnBackbufferSet(size_changed, is_first);
 }
 
 void Presenter::SetBackbuffer(SurfaceInfo info)
 {
+  const bool is_first = m_backbuffer_width == 0 && m_backbuffer_height == 0;
```
Comment (iwubcode) [inline Source/Core/VideoCommon/Present.cpp:219]: Isn't this implied by `SurfaceInfo` and the next statement?

Comment (iwubcode) [reply Source/Core/VideoCommon/Present.cpp:219]: Ah, I see, you need to call it out explicitly, nm

File: Source/Core/VideoCommon/Present.cpp [hunk 3]
```diff
@@ -201,20 +204,42 @@ void Presenter::ProcessFrameDumping(u64 ticks) const
...
+
+void Presenter::OnBackbufferSet(bool size_changed, bool is_first_set)
+{
   UpdateDrawRectangle();
+
+  // Automatically update the resolution scale if the window size changed,
+  // or if the game XFB resolution changed.
+  if (size_changed && !is_first_set && g_ActiveConfig.iEFBScale == EFB_SCALE_AUTO_INTEGRAL &&
+      m_auto_resolution_scale != AutoIntegralScale())
+    g_framebuffer_manager->RecreateEFBFramebuffer();
```
Comment (iwubcode) [inline Source/Core/VideoCommon/Present.cpp:240]: Not sure how I feel about this.  The framebuffer manager has a dependency on the presenter and the presenter has a dependency on the frame buffer manager.

Maybe the framebuffer manager could react to presenter back buffer changes using a callback or an event?

File: Source/Core/VideoCommon/Present.cpp (hunk 1 above)
Comment (iwubcode) [reply Source/Core/VideoCommon/Present.cpp:238]: Why was this marked as resolved?  You're not following the style guide here.

File: Source/Core/VideoCommon/Present.cpp (hunk 3 above)
Comment (Filoppi) [reply Source/Core/VideoCommon/Present.cpp:241]: I have considered a couple alternatives, included the ones discussed on Discord, but I'm not happy about any of them at the moment, they add a lot of boilerplate code to fix nothing really IMO. This can easily be patched up later if there was ever any need to.

**Comment by Filoppi** [comment]
I have fixed the FIFOCI rare failures, it was caused by `m_last_xfb_width` or `m_last_xfb_height` being zero sometimes.

**Comment by JMC47** [comment]
If there's no comments made on this soon, I'm going to assume everyone who wants to look at this has looked at it unless someone says otherwise.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp [hunk 4]
```diff
@@ -553,8 +553,8 @@ void EnhancementsWidget::AddDescriptions()
       "causes slowdowns or graphical issues.<br><br><dolphin_emphasis>If unsure, leave "
       "this unchecked.</dolphin_emphasis>");
   static const char TR_WIDESCREEN_HACK_DESCRIPTION[] = QT_TR_NOOP(
-      "Forces the game to output graphics for any aspect ratio. Use with \"Aspect Ratio\" set to "
-      "\"Force 16:9\" to force 4:3-only games to run at 16:9.<br><br>Rarely produces good "
+      "Forces the game to output graphics for any aspect ratio. (e.g.) Use with \"Aspect Ratio\" "
+      "set to \"Force 16:9\" to force 4:3-only games to run at 16:9.<br><br>Rarely produces good "
```
Comment (JosJuice) [inline Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:557]: @MayImilae, any comments about the working here? Mainly thinking about the way "e.g." is used.

File: Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp [hunk 5]
```diff
@@ -207,11 +238,15 @@ void GeneralWidget::AddDescriptions()
...
                  "this unchecked.</dolphin_emphasis>");
-  static const char TR_ASPECT_RATIO_DESCRIPTION[] = QT_TR_NOOP(
-      "Selects which aspect ratio to use when rendering.<br><br>Auto: Uses the native aspect "
-      "ratio<br>Force 16:9: Mimics an analog TV with a widescreen aspect ratio.<br>Force 4:3: "
-      "Mimics a standard 4:3 analog TV.<br>Stretch to Window: Stretches the picture to the "
-      "window size.<br><br><dolphin_emphasis>If unsure, select Auto.</dolphin_emphasis>");
+  static const char TR_ASPECT_RATIO_DESCRIPTION[] =
+      QT_TR_NOOP("Selects which aspect ratio to use when rendering.<br>"
+                 "Each game can have a slightly different default aspect ratio."
+                 "<br><br>Auto: Uses the native aspect ratio"
```
Comment (JosJuice) [inline Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp:244]: Is there a reason why what I assume is the same concept is first referred to as "default aspect ratio" and then "native aspect ratio"?

File: Source/Core/VideoCommon/VideoConfig.h [hunk 6]
```diff
@@ -21,9 +21,10 @@ constexpr int EFB_SCALE_AUTO_INTEGRAL = 0;
...
 enum class AspectMode : int
 {
-  Auto,
-  AnalogWide,
-  Analog,
+  Auto,           // 4:3 or 16:9
+  ForceWide,      // 16:9
+  ForceStandard,  // 4:3
+  Custom,
   Stretch,
```
Comment (JosJuice) [inline Source/Core/VideoCommon/VideoConfig.h:28]: Please don't reorder enum values that are saved to settings. This will cause everyone who previously had Stretch selected to now have Custom selected.

File: Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp [hunk 7]
```diff
@@ -138,13 +154,28 @@ void GeneralWidget::ConnectWidgets()
     Config::SetBaseOrCurrent(Config::GFX_ADAPTER, index);
     emit BackendChanged(QString::fromStdString(Config::Get(Config::MAIN_GFX_BACKEND)));
   });
+  connect(m_aspect_combo, qOverload<int>(&QComboBox::currentIndexChanged), this, [&](int index) {
+    const bool is_custom_aspect_ratio = (index == (int)AspectMode::Custom);
```
Comment (JosJuice) [inline Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp:158]:
```suggestion
    const bool is_custom_aspect_ratio = (index == static_cast<int>(AspectMode::Custom));
```

File: Source/Core/VideoCommon/Present.cpp [hunk 8]
```diff
@@ -365,38 +401,28 @@ void* Presenter::GetNewSurfaceHandle()
...
-  u32 width = EFB_WIDTH * target_width / m_last_xfb_width;
-  u32 height = EFB_HEIGHT * target_height / m_last_xfb_height;
-  return std::max((width - 1) / EFB_WIDTH + 1, (height - 1) / EFB_HEIGHT + 1);
+    source_height = std::round(source_width / target_aspect_ratio);
+  const u32 width_scale =
+      source_width > 0 ? ((target_width + (source_width - 1)) / source_width) : 1;
+  const u32 height_scale =
+      source_height > 0 ? ((target_height + (source_height - 1)) / source_height) : 1;
+  // Limit to the max to avoid creating textures larger than their max supported resolution.
+  return std::min(std::max(width_scale, height_scale), (u32)Config::Get(Config::GFX_MAX_EFB_SCALE));
```
Comment (JosJuice) [inline Source/Core/VideoCommon/Present.cpp:422]:
```suggestion
  return std::min(std::max(width_scale, height_scale), static_cast<u32>(Config::Get(Config::GFX_MAX_EFB_SCALE)));
```

**Review by JosJuice — COMMENTED** [review]
Regarding the commit "Video: fix auto resolution scale calculations": What was wrong with the calculations? And the description of this commit probably shouldn't be "asd".

File: Source/Core/VideoCommon/VideoConfig.h (hunk 6 above)
Comment (Filoppi) [reply Source/Core/VideoCommon/VideoConfig.h:28]: I was aware of that, I didn't think it mattered much as it's probably just a few people that would actually be the ones that'd be interested in this feature.
I can move `Custom` after stretch though.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 4 above)
Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:557]: The point here is to highlight that this feature isn't exclusively made for the "Force 16:9" aspect ratio setting, it works with "Stretch" and "Custom" aspect ratios as well.

Comment (JosJuice) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:565]: Yes, I don't think it's a bad thing. I just found the way of sticking "e.g." in brackets to be a bit odd.

File: Source/Core/Core/HW/VideoInterface.cpp [hunk 9]
```diff
@@ -477,7 +477,7 @@ float VideoInterfaceManager::GetAspectRatio() const
   // signal (which would be 4:3)
 
   // This function only deals with standard aspect ratios. For widescreen aspect ratios,
-  // multiply the result by 1.33333..
+  // multiply the result by 1.33333 (the ratio between 16:9 and 4:3)
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/HW/VideoInterface.cpp:480]: The `..` was intentional here to suggest 'infinitely repeating', I'm pretty sure.

File: Source/Core/VideoCommon/Present.cpp [hunk 10]
```diff
@@ -292,15 +293,22 @@ float Presenter::CalculateDrawAspectRatio(bool allow_stretch) const
...
       (aspect_mode == AspectMode::Auto && g_widescreen->IsGameWidescreen()))
   {
-    return AspectToWidescreen(aspect_ratio);
+    return SourceAspectRatioToWidescreen(source_aspect_ratio);
+  }
+  // For the "custom" mode, we force the exact target aspect ratio, without
+  // acknowleding the difference between the source aspect ratio and 4:3.
+  else if (aspect_mode == AspectMode::Custom)
+  {
+    return g_ActiveConfig.GetCustomAspectRatio();
```
Comment (AdmiralCurtiss) [inline Source/Core/VideoCommon/Present.cpp:336]: I haven't tested this but I suspect just ignoring the VI-timing derived aspect ratio will negatively affect games that use a nonstandard timing -- what immediately comes to mind is those cinema-style widescreen scenes in Rouge Squadron. While I agree it's useful to be able to set the desired aspect ratio, I'm wondering if this shouldn't be multiplied onto the VI timings instead of completely overriding them.

Comment (AdmiralCurtiss) [reply Source/Core/VideoCommon/Present.cpp:336]: Yeah, exactly what I expected:

Using the 'Force 4:3' option: https://github.com/dolphin-emu/dolphin/assets/4522237/5cd95da0-137c-421f-972f-7914cd28efb7
Using the 'Custom' option and entering 4/3: https://github.com/dolphin-emu/dolphin/assets/4522237/248456f4-b876-4103-ab50-61af05d75c37

Despite both being '4:3', they are different, because the former respects the VI timings and the latter doesn't. This is especially awkward if the game switches VI timings between eg. cutscenes and gameplay or whatever -- as this game does.

I see multiple options here:
- Instead of treating the custom aspect ratio as an absolute, treat it as a modifier onto the VI aspect ratio. Calculate the difference to the VI-timing provided 4:3 and multiply onto that -- note that this is exactly what the current 16:9 option does. Probably the cleanest solution, but means that the resulting image is not actually exactly the entered aspect ratio in most cases.
- Just ignore this and stick with the absolute aspect ratio and hope the user knows what they're doing. Note that this will make the Custom option somewhat incompatible with games like Rogue Leader here.
- Or perhaps have both of these options? Might just confuse users though.

File: Source/Core/VideoCommon/VideoConfig.h (hunk 6 above)
Comment (AdmiralCurtiss) [reply Source/Core/VideoCommon/VideoConfig.h:27]: You'll need to swap the labels in the UI now too, they're wrong in the current version of the PR.

File: Source/Core/Core/HW/VideoInterface.cpp (hunk 9 above)
Comment (Filoppi) [reply Source/Core/Core/HW/VideoInterface.cpp:480]: Ah 😅. I'll fix that.

File: Source/Core/VideoCommon/Present.cpp (hunk 10 above)
Comment (Filoppi) [reply Source/Core/VideoCommon/Present.cpp:336]: Wouldn't the 'Stretch' setting in a 4:3 act like 'Custom' set to 4:3? That was the main intention here. It's just to allow specific "stretch" values without actually needing to resize the window to that specific desired aspect ratio (which sometimes is near impossible, e.g. I have a 32:9 screen and use 21:9 game mods).
I'd be fine with somebody else adding a 'Stretch (relative)' setting later on which acts as a relative aspect ratio modifier (ratio), I think it's beyond the scope of this PR and would like to thing about it after.
This is stable for now and works for the intent it has. It's a very specific use case and already helps users deal with specific scenarios IMO.

Comment (AdmiralCurtiss) [reply Source/Core/VideoCommon/Present.cpp:336]: > Wouldn't the 'Stretch' setting in a 4:3 act like 'Custom' set to 4:3? (Maybe not).

Assuming you mean 'in a 4:3 window', yes that would be the same thing. If this is purely supposed to be a replacement for Stretch without having to fiddle the window into the correct size, then yes, this is implemented correctly as far as I can tell. In that case I would like to make that more obvious in the UI in some way though.

Comment (AdmiralCurtiss) [reply Source/Core/VideoCommon/Present.cpp:336]: Maybe just rename the option to "Stretch (Custom)"?

Comment (Filoppi) [reply Source/Core/VideoCommon/Present.cpp:336]: I got the code for the two "Custom - Absolute" and "Custom - Relative" modes ready, we can make a follow up PR

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 4 above)
Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:565]: any suggestions?

File: Source/Core/VideoCommon/Present.cpp (hunk 10 above)
Comment (AdmiralCurtiss) [reply Source/Core/VideoCommon/Present.cpp:336]: We discussed this at length in Discord, but to immortalize it here: Yes, a follow-up sounds fine to me.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 4 above)
Comment (MayImilae) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:565]: > Forces the game to output graphics for any aspect ratio. (e.g.) Use with "Aspect Ratio" set to "Force 16:9" to force 4:3-only games to run at 16:9.

Honestly I think that example makes things more confusing, rather than helping. Like, the bit that explains it is the start "force a game to output graphics for any aspect ratio" is how it works. If we want to explain applications and usage for it though, we'd have to write a paragraph talking about how it increases FOV and things. Saying it can force 4:3 games to 16:9 isn't wrong but it isn't making it obvious why you'd want to use it. 

I'm also not happy with a few other bits of the language here, such as "for any aspect ratio" (should be at), "rarely produces good results" (it's more 50-50 in my experience, not great but hardly "rare"), and it's focus on widescreen when it can do any aspect ratio from ultrawide to portrait. Let me give you an alternate take for this that may be better.

>Forces the game to output graphics at any aspect ratio without stretching the image by expanding the view frustum. This is a hack, and its results will vary widely game to game. Game-specific AR/Gecko-code aspect ratio patches will out perform this hack, use them instead whenever possible.

File: Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp [hunk 11]
```diff
@@ -207,11 +238,15 @@ void GeneralWidget::AddDescriptions()
...
-      "Selects which aspect ratio to use when rendering.<br><br>Auto: Uses the native aspect "
-      "ratio<br>Force 16:9: Mimics an analog TV with a widescreen aspect ratio.<br>Force 4:3: "
-      "Mimics a standard 4:3 analog TV.<br>Stretch to Window: Stretches the picture to the "
-      "window size.<br><br><dolphin_emphasis>If unsure, select Auto.</dolphin_emphasis>");
+  static const char TR_ASPECT_RATIO_DESCRIPTION[] =
+      QT_TR_NOOP("Selects which aspect ratio to use when rendering.<br>"
+                 "Each game can have a slightly different native aspect ratio."
+                 "<br><br>Auto: Uses the native aspect ratio"
+                 "<br>Force 16:9: Mimics an analog TV with a widescreen aspect ratio."
+                 "<br>Force 4:3: Mimics a standard 4:3 analog TV."
```
Comment (MayImilae) [inline Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp:246]: I know this is an old bit here, but while we are at it... the use of "Analog TV" here bugs me. Like, I get why that's there, it's assuming a "dumb" display that defaults to stretching, but that's not necessarily true even in analog.  Even the use of "standard" is silly. I'd much prefer:

>Force 4:3: Mimics a 4:3 display.
>Force 16:9 Mimics a display with a 16:9 widescreen aspect ratio.

That's all this needs really. 

Like, that's all this needs.

File: Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp [hunk 12]
```diff
@@ -207,11 +238,15 @@ void GeneralWidget::AddDescriptions()
...
                  "a separate render window.<br><br><dolphin_emphasis>If unsure, leave "
                  "this unchecked.</dolphin_emphasis>");
-  static const char TR_ASPECT_RATIO_DESCRIPTION[] = QT_TR_NOOP(
-      "Selects which aspect ratio to use when rendering.<br><br>Auto: Uses the native aspect "
-      "ratio<br>Force 16:9: Mimics an analog TV with a widescreen aspect ratio.<br>Force 4:3: "
-      "Mimics a standard 4:3 analog TV.<br>Stretch to Window: Stretches the picture to the "
-      "window size.<br><br><dolphin_emphasis>If unsure, select Auto.</dolphin_emphasis>");
+  static const char TR_ASPECT_RATIO_DESCRIPTION[] =
+      QT_TR_NOOP("Selects which aspect ratio to use when rendering.<br>"
+                 "Each game can have a slightly different native aspect ratio."
```
Comment (MayImilae) [inline Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp:243]: That is a very good addition. We've needed this here for a long time. 👍🏻

**Review by MayImilae — COMMENTED** [review]
Gave this a review of the user facing strings.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 4 above)
Comment (mbc07) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:565]: As someone who doesn't have English as my main language and that will have to deal with translating this later, what a "view frustum" is supposed to mean in that context? I looked up in a few dictionaries but I'm still having a hard time correlating the meaning of that expression with what this setting actually does...

Comment (MayImilae) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:565]: It comes up pretty quickly with a Google. https://en.wikipedia.org/wiki/Viewing_frustum

It's the technical term for what is within the pyramid of the camera's field of view. We could avoid the term but it would be a bit wordy to explain without it, plus it's pretty easy to google. It's also where the term "frustum culling" comes from, which is a parlance that is reasonably well known in gaming culture... at least in English native countries.

File: Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp (hunk 11 above)
Comment (AdmiralCurtiss) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp:246]: I know what you want to say but I don't think "Ignores the game's configured aspect ratio" is quite right here either. Both of these options *are* taking the game's video configuration into account, they just ignore our heuristic to decide whether the game is trying to render 4:3 or 16:9.

Comment (AdmiralCurtiss) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp:246]: (this stuff is really hard to put into a short single-line string for users who don't have technical knowledge so I suppose it's okay to be wrong in order the get the point across, but yeah...)

Comment (MayImilae) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp:246]: Doesn't Force 4:3 and Force 16:9 outright stretch to exactly 4:3 and 16:9?

Comment (AdmiralCurtiss) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp:246]: Yeah they don't, they behave exactly like 'Auto' except it skips the heuristic and just assumes its result.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 4 above)
Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:565]: Ok, the new wording is much better, I'll change to that.

File: Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp (hunk 11 above)
Comment (Filoppi) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp:246]: Yes, as much as I don't like the old wording, it kind of makes sense. If a game was meant for 4:3 output but played on a 16:9 CRT TV in 16:9 mode, the same distortion would now apply in Dolphin when you force 16:9. The opposite for 4:3 being forced on 16:9 games. Overall, these options are barely ever useful if not for people that use CTRs? Or for games that fail the heuristic aspect ratio detection I guess.
One thing that is confusing about their wording is that it's not exactly 16:9, at least not most of the times. So maybe they could be called `Force ~16:9` or something like that :D.
Though we can also fix these later without it blocking this PR, given I didn't really change these lines myself.

Comment (AdmiralCurtiss) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp:246]: Yeah... I'm thinking that this will need tweaking but at the same time I don't think it makes sense to hold up this PR over this. Go fix the other string and then after this is merged I'll open up a new PR for further discussion on the phrasing of this one, if that's fine with everyone.
